src/mono/mono/mini/method-to-ir.c

```diff
@@ -2303,6 +2303,18 @@ emit_not_supported_failure (MonoCompile *cfg)
 	mono_emit_jit_icall (cfg, mono_throw_not_supported, NULL);
 }
 
+static void
+emit_type_load_failure (MonoCompile* cfg, MonoClass* klass)
+{
+	MonoClass* failed_class = klass;
+	if (G_UNLIKELY (!failed_class)) 
+		failed_class = mono_defaults.object_class;
+
+	MonoInst* iargs[1];
+	EMIT_NEW_CLASSCONST (cfg, iargs [0], failed_class);
```

Comment:
I'm not sure about blaming `System.Object` for failures where we don't have more context.  you can emit a null pointer AOT const and check for it in `mono_throw_type_load` and then use some more generic error message.